src/Server/TCPHandler.cpp

```diff
@@ -1252,7 +1252,7 @@ void TCPHandler::receiveQuery()
         original_session_client_info.current_query_id.clear();
         original_session_client_info.current_address = {};
 
-        session = std::make_unique<Session>(server.context(), ClientInfo::Interface::TCP_INTERSERVER);
+        session = std::make_unique<Session>(global_context, ClientInfo::Interface::TCP_INTERSERVER);
```

Comment:
`TCPHandler` will depend on the global context anyway no matter how it's written. And during shutdown we must do the following two things:

1. Check if some connections are still alive and give them time to finish ([link](https://github.com/ClickHouse/ClickHouse/blob/c3f177f0ac0726424d3ecee1d444ae9045438e5b/programs/server/Server.cpp#L1748))
2. Ensure than nobody is going to use the global context after calling `Context::shutdown()`.

With unbinding connections from servers it seems your code is going to break both rules.

Reply:
You are right, I've updated the patch to remember all existing servers.